builtin/credential/radius/path_login.go

```diff
@@ -0,0 +1,174 @@
+	}
+}
+
+func (b *backend) pathLogin(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	username := d.Get("username").(string)
+	password := d.Get("password").(string)
+
+	policies, resp, err := b.RadiusLogin(req, username, password)
+	// Handle an internal error
+	if err != nil {
+		return nil, err
+	}
+	if resp != nil {
+		// Handle a logical error
+		if resp.IsError() {
+			return resp, nil
+		}
+	}
+
+	resp.Auth = &logical.Auth{
+		Policies: policies,
+		Metadata: map[string]string{
+			"username": username,
+			"policies": strings.Join(policies, ","),
+		},
+		InternalData: map[string]interface{}{
+			"password": password,
+		},
+		DisplayName: username,
+		LeaseOptions: logical.LeaseOptions{
+			Renewable: true,
+		},
+	}
+	return resp, nil
+}
+
+func (b *backend) pathLoginRenew(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	var err error
+
+	cfg, err := b.Config(req)
+	if err != nil {
+		return nil, err
+	}
+
+	username := req.Auth.Metadata["username"]
+	password := req.Auth.InternalData["password"].(string)
+
+	var resp *logical.Response
+	var loginPolicies []string
+	var user *UserEntry
+	if cfg.ReauthOnRenew {
+		loginPolicies, resp, err = b.RadiusLogin(req, username, password)
+
+		if err != nil || (resp != nil && resp.IsError()) {
+			return resp, err
+		}
+
+	} else {
+		user, err = b.user(req.Storage, username)
+		if err != nil {
+			return nil, err
+		}
+		if user == nil {
+			// User no longer exists, do not renew
+			return nil, fmt.Errorf("Renew not allowed for unknown users")
+		}
+		loginPolicies = user.Policies
+	}
+
+	if !policyutil.EquivalentPolicies(loginPolicies, req.Auth.Policies) {
+		return nil, fmt.Errorf("policies have changed, not renewing")
+	}
+
+	return framework.LeaseExtend(0, 0, b.System())(req, d)
+}
+
+func (b *backend) RadiusLogin(req *logical.Request, username string, password string) ([]string, *logical.Response, error) {
+
+	cfg, err := b.Config(req)
+	if err != nil {
+		return nil, nil, err
+	}
+	if cfg == nil || cfg.Host == "" || cfg.Secret == "" {
+		return nil, logical.ErrorResponse("radius backend not configured"), nil
+	}
+
+	hostport := net.JoinHostPort(cfg.Host, strconv.Itoa(cfg.Port))
+
+	packet := radius.New(radius.CodeAccessRequest, []byte(cfg.Secret))
+	packet.Add("User-Name", username)
+	packet.Add("User-Password", password)
+	packet.Add("NAS-Port", uint32(cfg.NasPort))
+
+	client := radius.Client{
+		DialTimeout: time.Duration(cfg.DialTimeout) * time.Second,
+		ReadTimeout: time.Duration(cfg.ReadTimeout) * time.Second,
+	}
+
+	received, err := client.Exchange(packet, hostport)
+
+	if err != nil {
+		return nil, logical.ErrorResponse(err.Error()), nil
+	}
+
+	if received.Code != radius.CodeAccessAccept {
+		return nil, logical.ErrorResponse("access denied by the authentication server"), nil
+	}
+
+	radiusResponse := &logical.Response{}
+
+	// Retrieve policies
+	var policies []string
+	user, err := b.user(req.Storage, username)
```

Comment:
Do you reckon it's worth having a config entry where a list of policies are granted upon successful RADIUS auth instead of forcing it to be `default` and relying on a code hack?